feat: add `-a` flag to `gh run list`

joshuajtward opened the pull request:
<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->
Fixes #9038

This PR adds an `--all` (or `-a`) flag to `gh run list` so that users can get results for disabled workflows when searching for runs via the `workflow_name`. This mirrors the way the `gh workflow list -a` works. For more detail please see #9038

joshuajtward committed:
feat: add `-a` flag to `gh run list`

andyfeller commented on pkg/cmd/run/list/list.go:
```diff
@@ -111,6 +121,9 @@ func listRun(opts *ListOptions) error {
 	opts.IO.StartProgressIndicator()
 	if opts.WorkflowSelector != "" {
 		states := []workflowShared.WorkflowState{workflowShared.Active}
+		if opts.All {
+			states = append(states, workflowShared.DisabledManually, workflowShared.DisabledInactivity)
+		}
```
**@williammartin : any way you can think of an alternative to this that are not impacted by adding new workflow states?**

With `ResolveWorkflow` and other places designed to have calling code pass in all states interested in, this logic is a bit brittle when someone wants all workflows:

https://github.com/cli/cli/blob/a10e5328982f72a2098deb7c32021beca7ab2ff5/pkg/cmd/workflow/shared/shared.go#L185-L207

So this can go 1 of 2 ways in my mind:

1. Every place calling the same logic has to be kept in sync with various workflow states like this
2. Add a new workflow state used for retrieval, essentially an `any` option

https://github.com/cli/cli/blob/a10e5328982f72a2098deb7c32021beca7ab2ff5/pkg/cmd/workflow/shared/shared.go#L22-L26

This would essentially have underlying code checking for `any` or the specified state:

https://github.com/cli/cli/blob/a10e5328982f72a2098deb7c32021beca7ab2ff5/pkg/cmd/workflow/shared/shared.go#L100-L111

https://github.com/cli/cli/blob/a10e5328982f72a2098deb7c32021beca7ab2ff5/pkg/cmd/workflow/shared/shared.go#L163-L183

andyfeller replied:
@joshuajtward : If we can't find a better solution, a middle ground might be adding a comment within the shared workflow logic to call out places like this that might be updated as new states are added.

joshuajtward replied:
@andyfeller I've added a comment calling out this potential pitfall. Your solution of an `any` method is nice, but I think it may be beyond the scope of this PR, would you want me to raise it as a separate issue? 

I had a quick look for ways to do something similar to `splat` in Python or spread in JS, but I'm pretty new to Go so it felt like too much of a rabbit hole for this simple case...

williammartin replied:
I guess my first question would be "Is there any precedent for new states being added here"? Seems like it should be fairly static.

Adding a new `WorkflowState` enumeration of `Any` has its own potential to cause confusion because it implies that a `Workflow` can have that state.

I don't feel particularly strongly either way though. If you pushed me I'd say leave it as is with a comment.

andyfeller replied:
Thanks for the feedback, @williammartin!  I've gone ahead and wrote some tests to ensure this works as expected though I'm realizing there are more test cases to capture here as none of them deal with disabled workflows and no workflow selector. 😞

joshuajtward committed:
add comment to call out potentially brittle use of workflowShared

andyfeller committed:
Add `gh run list` tests for workflow selector

This commit adds tests to ensure disabled workflows are outputted as expected when called by name and the appropriate flags.

andyfeller reviewed (approved):
Thank you for your patience, @joshuajtward; I think this looks as good as we're likely going to get it along with some tests to make sure it works as expected 👍
